### PR TITLE
[Draft] API Definition review

### PR DESCRIPTION
**Pull request by jufantozzi**
**Pull Request check list**

- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

**Description of change**

This PR is a draft to promote discussion around updating our API definitions. 
The goal is to review the spec so it aligns with latest definition we have while also reflecting goals for our PoC.
 

**Which issue this PR fixes**

### CODE REVIEW

**Comment by jufantozzi** [comment]
Hey @thisab, just realized that William Souza does not have access to this repo. Would you please share his user so I can add him to this discussion?

**Comment by thisab** [comment]
> Hey @thisab, just realized that William Souza does not have access to this repo. Would you please share his user so I can add him to this discussion?

Sure, here it is: https://github.com/wiliamsouza

**Comment by wiliamsouza** [comment]
I whould like to discuss the use of `HTTP` method `PUT` on API definition. Should we keep it or change to use `PATCH`?
`PATCH` is for making partial changes to an existing resource other hand `PUT` is used for replace an existing resource content[1]. 

[1] https://josipmisko.com/posts/patch-vs-put-rest-api

File: spec/api/management.yaml [hunk 1]
```diff
@@ -14,29 +14,30 @@ components:
...
+#        - name
     FederationGroup:
-      # A Federation Group is a group of SpireServers (representation of the bridge)
+      # A Federation Group represents a bridge for its Members,
+      # allowing relationships to be established
       type: object
       properties:
         id:
           type: integer
           format: int64
```
Comment (thisab) [inline spec/api/management.yaml:37]: @jufantozzi could we use format: uuid to store the ID?

Comment (thisab) [inline spec/api/management.yaml:37]:
```suggestion
          type: string
          format: uuid
```
@jufantozzi maybe we could use format: uuid to store the ID?

**Review by thisab — COMMENTED** [review]
Adding my 2 cents on a topic discussed earlier.

- There are some other API definitions (/organizations) that we may strip for the PoC?
- Will federationGroupMemberships become just Memberships ?

**Comment by mchurichi** [comment]
> I whould like to discuss the use of `HTTP` method `PUT` on API definition. Should we keep it or change to use `PATCH`? `PATCH` is for making partial changes to an existing resource other hand `PUT` is used for replace an existing resource content[1].
> 
> [1] https://josipmisko.com/posts/patch-vs-put-rest-api

PATCH is what people want 99% of the time, I'd suggest to use it instead of PUT

**Comment by jufantozzi** [comment]
> I wasn't aware we are stripping Organization. Is it for PoC purposes?

AFAIK by stripping I think he meant we're removing.

**Comment by mchurichi** [comment]
> > I wasn't aware we are stripping Organization. Is it for PoC purposes?
> 
> AFAIK by stripping I think he meant we're removing. Was your question regarding it's removal?

Got it, and I wasn't aware of that decision.

More food for thought:
- `/federationRelationships` -> `/relationships`
- `/spireServers`  -> `/members`

**Comment by thisab** [comment]
> > I wasn't aware we are stripping Organization. Is it for PoC purposes?
> 
> AFAIK by stripping I think he meant we're removing. Was your question regarding it's removal?

Edited my previous comment, but yes and for PoC purposes only.
Maybe: `/federationGroupMemberships` -> `/membership`

File: spec/api/management.yaml (hunk 1 above)
Comment (mchurichi) [reply spec/api/management.yaml:37]: how does uuid's translates to the different dbms?

Comment (thisab) [reply spec/api/management.yaml:37]: I can say that [gorm](https://medium.com/@the.hasham.ali/how-to-use-uuid-key-type-with-gorm-cc00d4ec7100) supports it. [MongoDB](https://www.mongodb.com/docs/manual/reference/method/UUID/) seems too

**Comment by jufantozzi** [comment]
Made the updates already agreed on here. There one concern that came to mind when reviewing, and this is something I also asked on the design doc: How will the trust bundle endpoint work? will we actually allow for those to be manually manipulated?

**Comment by mchurichi** [comment]
> Made the updates already agreed on here. There's one concern that came to mind when reviewing, and this is something I also asked on the design doc: How will the trust bundle endpoint work? will we actually allow for those to be manually manipulated?

What do you mean by manually manipulated? To allow an admin user to change/delete Trust Bundles? that doesn't feel quite right.

AFAIK the spire server does not provide a method to change/update its own current bundle. You can only append a new one, but not delete or change one that is already there.

**Comment by jufantozzi** [comment]
> What do you mean by manually manipulated? To allow an admin user to change/delete Trust Bundles? that doesn't feel quite right.

I'm referring to this [very hidden path](https://github.com/HewlettPackard/galadriel/pull/43/files#diff-df80259918c1ba12005e4825cf76acd82233f079d30c7e88d84489bd7acc81fcR727), I'm not sure either what was the idea behind it but seems off to me.

> AFAIK the spire server does not provide a method to change/update its own current bundle. You can only append a new one, but not delete or change one that is already there.

that's my understanding as well.

**Comment by mchurichi** [comment]
> > What do you mean by manually manipulated? To allow an admin user to change/delete Trust Bundles? that doesn't feel quite right.
> 
> I'm referring to this [very hidden path](https://github.com/HewlettPackard/galadriel/pull/43/files#diff-df80259918c1ba12005e4825cf76acd82233f079d30c7e88d84489bd7acc81fcR727), I'm not sure either what was the idea behind it but seems off to me.

IMO that shouldn't be allowed.

**Comment by jufantozzi** [comment]
Closing this for now, we should be re-visiting in the future

**Pull request closed without merging**